v3io/dataplane/transport/httpclient.py

```diff
@@ -102,18 +127,40 @@ def wait_response(self, request, raise_for_status=None, num_retries=1):
 
                 response = v3io.dataplane.response.Response(request.output, status_code, headers, response_body)
 
-                self._free_connections.put(connection, block=True)
-
+                # Return connection to pool if successful
+                with self._lock:
+                    if not self._closed:
+                        try:
+                            self._free_connections.put(connection, block=False)
+                        except Exception as e:
+                            self._logger.warn_with(
+                                "Failed to return connection to pool", exception=str(e), connection_id=id(connection)
+                            )
+                            connection.close()
                 response.raise_for_status(request.raise_for_status or raise_for_status)
-
                 return response
 
             except v3io.dataplane.response.HttpResponseError as response_error:
                 self._logger.warn_with(f"Response error: {response_error}")
+                # Return connection to pool even on HTTP errors, as the connection is still valid
+                with self._lock:
+                    if not self._closed:
+                        try:
+                            self._free_connections.put(connection, block=False)
+                        except Exception as e:
+                            self._logger.warn_with(
+                                "Failed to return connection to pool", exception=str(e), connection_id=id(connection)
+                            )
+                            connection.close()
```

Comment:
As discussed, this change would introduce a connection leak rather than fixing one.